### PR TITLE
:sparkles: Add cluster owner references to external MS/MD objects

### LINKED ISSUE #1775: Delete cluster doesn't delete awsmachinetemplate
**What steps did you take and what happened:**
Create a cluster API cluster
Create some machine deployment with awsmachinetemplate
Delete a cluster API cluster
The awsmachinetemplate is not deleted

The issue is, if I then create a new cluster with the same cluster name and same machine deployment name, it would fail as it will try to use the old awsmachinetemplate from the old cluster that wasn't deleted. For developer cluster, the awsmachinetemplate contain info such as vm flavor and ssh key which would be using the wrong one. For HA cluster, it would contain subnet ID, which will be previous cluster's previous subnet. The subnet would have been deleted in the previous cluster delete. Thus causing failure saying "subnet not exist"

**What did you expect to happen:**
When deleting a cluster, the template should all be deleted as part of the cluster. The template should have an owner reference.

**Anything else you would like to add:**

AWSmachinetemplate data being reused if the second cluster and machine deployment has the same name. 
ami: {}
      iamInstanceProfile: nodes.tmc.cloud.vmware.com
      instanceType: m5.xlarge
      rootDeviceSize: 50
      sshKeyName: olympus-default
      subnet:
        id: subnet-0df10de05adf6addf

**Environment:**

- Cluster-api version: v1alpha2 cluster

/kind bug

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1775
ref #1718

### CODE REVIEW

File: controllers/machineset_controller.go [hunk 1]
```diff
@@ -153,6 +154,17 @@ func (r *MachineSetReconciler) reconcile(ctx context.Context, machineSet *cluste
 		}
 	}
 
+	// Make sure to reconcile the external infrastructure reference.
+	if err := r.reconcileExternalReference(ctx, cluster, machineSet.Spec.Template.Spec.InfrastructureRef); err != nil {
+		return ctrl.Result{}, err
+	}
+	// Make sure to reconcile the external bootstrap reference, if any.
+	if machineSet.Spec.Template.Spec.Bootstrap.ConfigRef != nil {
+		if err := r.reconcileExternalReference(ctx, cluster, machineSet.Spec.Template.Spec.InfrastructureRef); err != nil {
```
Comment (ncdc) [inline controllers/machineset_controller.go]:
```suggestion
		if err := r.reconcileExternalReference(ctx, cluster, machineSet.Spec.Template.Spec.Bootstrap.ConfigRef); err != nil {
```

File: controllers/machineset_controller_test.go [hunk 2]
```diff
@@ -131,6 +133,22 @@ var _ = Describe("MachineSet Reconciler", func() {
...
+
+			return util.HasOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
+				APIVersion: clusterv1.GroupVersion.String(),
+				Kind:       "Cluster",
+				Name:       testCluster.Name,
+				UID:        testCluster.UID,
+			})
+
+		}, timeout).Should(BeTrue())
+
```
Comment (ncdc) [inline controllers/machineset_controller_test.go]: Can you test the bootstrap ref too?

Comment (vincepri) [reply controllers/machineset_controller_test.go]: Done!

File: controllers/machineset_controller.go (hunk 1 above)
Comment (vincepri) [reply controllers/machineset_controller.go]: Done!

**Pull request merged**